### DISCUSSION TITLE
Black-oil properties for Oil & Gas applications

### DISCUSSION THREAD

**Discussion by Gualcar** [CoolProp/CoolProp#2297] [Ideas]
Hello everyone.

Do you think Black-oil correlations for Oil & Gas applications such multiphase-flow modeling, reservoir simulations, etc, is a good feature here in CoolProp?

With this feature I think CoolProp could be used by these people from the Petroleum Engineering. I don't know if this is the main idea of CoolProp to be a "hub" of fluid properties calculations, but I think it should be nice to have a unique library where you could perform black-oil or compositional analysis.

**Comment by ibell** [comment]
What is black oil?

**Comment by Gualcar** [comment]
It is a simplified model that consider hypothetically a mixture of oil and gas (maybe water) to describe fluids properties. It is extensively used in the Petroleum industry, because it does not require a lot of informations from the fluid, such the composition of each component.

To describe a fluid using black-oil model we require basic inputs such the solubility ratio and bubble point pressure (in a specified temperature), and as a result we can get the fluid density and viscosity, for example. These models are very simple, the pro is that it does not require computational power, but the con is that they are not so powerful as the compositional model (cubic EOS, for example). Also, they are mainly experimental correlations, which do not give us a thermodynamic approach such as EOSs.

I'm trying to compile different correlations to put them together. These models can be used in Reservoir simulators or Multiphase flow simulators to enhance the numerical solver performance, because in these simulators we need to calculate fluid properties in each control volume in the discretized domain.

I hope I explained a little what it is.

Reference -> A paper from IntechOpen: PVT Properties of Black Crude Oil (https://www.intechopen.com/chapters/64816)

**Comment by ibell** [comment]
I am not entirely sure this is a good fit for CoolProp. It doesn't exactly fit into the CoolProp modeling methodology, and I think cubic EOS are a better fit.

**Reply by Gualcar** [reply]
Yeah, I agree with you. Since the CoolProp has a root in thermodynamics methodology it does not make sense to implement a bunch of experimental correlations.

By the way, I started to study the CoolProp code base, soon I will be able to help with coding.

Thank you @ibell!

**Comment by pkirkham** [comment]
@Gualcar I'm a petroleum engineer and I've been using CoolProp for about half a year now to. Black oil correlations are totally inappropriate for CoolProp. In fact, I would go as far as to say that unless you are doing a manual calculation as a sense check, these days you shouldn't be using correlations. For hydrocarbon mixtures cubic EOS are a better fit, particularly recent variants like EPPR78. That said, CoolProp is very useful for pure fluids like CO2, particularly when considering applications like CCS, which is what brought me to this project originally. CoolProp does also already have implementations for SRK and PR cubics which you might find interesting.

Now, specifically talking Black Oil vs Compositional... yes Black Oil is faster, but not because it uses correlations. Black Oil is faster because it uses pre-computed look up tables (LUT) e.g., PVTO and PVTG keywords in an Eclipse simulation deck. Those tables are actually usually generated using a cubic EOS (at least if you're doing it properly). You can use correlations, but as I said before, you shouldn't be doing that these days. By having a LUT you don't have to solve the cubic EOS flash calculation each and every time you want to find a value. You just look up an interpolated result from the table. You might be interested to know that CoolProp has a similar functionality as it can tabulate the EOS results and then use a bicubic spline lookup which speeds things up.

Compositional models, on the other hand, perform a flash calculation for each and every cell, on each and every time step. Bazillions of iterations. This is why they take a long time. So to try to reduce the computation times, engineers have been employing lumping schemes etc. to reduce the number of components which just strikes me as somewhat backwards. What's the point in using a more sophisticated approach if you are then going to simplify your input data just to make it work? I've often wondered if there wasn't a better approach. Take dry gas recycling for example, a prime candidate for compositional modelling. The injected dry gas composition is known. The in-situ wet gas composition is also known. So in any given cell, at any given time, there is a composition that will be close to X% original and (1 - X)% injected gas. X can be determined easily enough using Rv (or tracers if you must). So this means we could simply generate a series of LUTs for our PVT properties for the range of X from 0% to 100%. Our pseudo-compositional model should now run at a similar speed to a Black Oil model as it is just doing look up interpolations to get the fluid viscosities, expansion factors, dewpoints etc. However, it would have close to the improved sophistication as a compositional model, potentially better as you are not compromising the composition by lumping it -- you can pre-compute the LUTs using the full composition.

I was going to write a simulator to test this idea out, but then I discovered that [OPM Flow](https://opm-project.org/) has already implemented it using their PVTSOL keyword. I strongly encourage you to look into it. The only snag you'll run into is that, to my knowledge, you won't easily find tools to generate the PVTSOL keyword input needed.

I've had a personal hobby project going for a while now to implement cubic EOS, including EPPR78, in a tool that allows generation of PVT tables for ECLIPSE simulator etc. I'm close to releasing a version of this, which will in all likelihood initially be free to download and use. It's probably still a few months away. If you're interested I can let you know when it's out.

**Reply by Gualcar** [reply]
Nice to meet you @pkirkham. Thank you for the explanation.

I agree with you that Cubic EOSs have a huge power to describe these systems. But still nowadays people use Black Oil correlations to describe fluids, because they lack data, such the composition of the fluid. Software as PIPESIM and OLGA (owned by SLB, as ECLIPSE is) have a bunch of these correlations implemented in the software to calculate fluid properties such as density and viscosity. I think the bigest pro using these correlations is that it requires a few input data.

I'm not used to reservoir simulators, I'm used to flow assurance simulators such as PIPESIM and OLGA. Usually they use the same approach as you said, they perform fluid properties calculations through look up tables. But they also have the option to calculate these properties rigorously using EOSs and flash algorithms or calling Black Oil correlations directly for each pressure-temperature pair in each control volume.

This analysis on how to accelerate compositional calculations is very interesting. I have always concerned about the assumption of constant composition for the fluid table, this consideration of different tables with different compositions to interpolate then seems to make sense.

Speaking about your personal project, it is very useful not only for ECLIPSE users, but also for PIPESIM and OLGA users. These software also use tables, so the same backend could be used to generate them. If you need help finishing this project let me know.

Regards!

**Reply by pkirkham** [reply]
Yes, correlations still have their place. As you say, existing industry software incorporates many of them and it's not hard to implement them using your own language of choice or even Excel or a calculator. Maybe things are different in the flow assurance world, but for reservoir simulation, if you don't know enough about your fluid to have a compositional analysis available, then perhaps you shouldn't be running a simulation. Material balance methods will suffice. So whilst you can always generate your PVT look up tables using correlations, you should also be questioning whether this is the right approach. Just because you can do something, doesn't mean you should!

As for generating tables for other software packages like PIPESIM and OLGA, I'm sure this would be possible. Maybe even trivial (provided the format used is straighforward). It wouldn't surprise me though if the formats for even OLGA and ECLIPSE are different, despite being owned by the same company.

On a slightly different but related note, what was the main interest incorporating correlations into CoolProp? In my mind, one of the great strengths of CoolProp is that it is a library that has been wrapped and implemented in so many different languages/environments. My colleagues and I have used the Excel add-in for quick prototyping and I use the Python wrapper in Jupyter notebooks for similar purposes. When I'm ready to incorporate into a desktop application, I use the Java wrappers within my own code. Are you thinking that having a library of correlations that could be accessed from a variety of languages is useful? How would you use it?

From my own perspective, I already have a library with many of the existing correlations that I've written in Java and from time to time, in various algorithms I write, I will call these methods and it is immensely useful for that purpose. The thing is, I suspect that anyone else needing this functionality would just do what I did and write their own library. After all, most of the correlations are very simple and fast, and can be written in just a handful of lines of code.

**Comment by pkirkham** [comment]
@Gualcar I've recently released the first version of a tool that will do compositional PVT analysis specifically for oil and gas. We've found it quite useful in our company -- the process engineers have particularly liked the ability for modelling CO2 with impurities, and with a reservoir engineering hat on I've been using it to generate ECLIPSE keywords for gas-condensate.

The underlying code uses CoolProp for pure fluids (e.g., pure CO2 or CH4), but switches to my own implementation of cubic EOS for mixtures. I've followed the [E-PPR78](https://doi.org/10.1016/j.ijggc.2016.11.015) equation of state.

A video and documentation on the tool are here: [https://pkirkham.github.io/pyrus/pvt-fluid-properties-tool/](https://pkirkham.github.io/pyrus/pvt-fluid-properties-tool/)

The software itself has no cost. I am contributing this as a professional courtesy to other petroleum engineers. It can be downloaded here: [https://pkirkham.github.io/pyrus/](https://pkirkham.github.io/pyrus/). It will run on Windows, MacOSX and Linux and is a portable distribution so you don't even need to install it before running.

Two specific things we have discussed above:

1. _What to do if you don't have the composition and only high-level data?_ I have some plans to look at developing consistent compositions from the same high-level data that are used for correlations. It seems to me that the best correlation isn't a simple equation, but is actually a proxy composition derived from high-level data, and then use a proper equation of state to derive properties from the proxy composition.
2. _Tables for PIPESIM and OLGA_? The tool currently generates tables for ECLIPSE based on the provided composition. I can probably do something similar for other programs but would ideally need test data to verify results against. Do you have known input compositions and output tables that you would be willing to share?
